chain/zcash/address.go

```diff
@@ -4,13 +4,112 @@ import (
 	"bytes"
 	"crypto/sha256"
 	"errors"
+	"fmt"
 
 	"github.com/btcsuite/btcd/chaincfg"
 	"github.com/btcsuite/btcutil"
 	"github.com/btcsuite/btcutil/base58"
+	"github.com/renproject/multichain/api/address"
+	"github.com/renproject/pack"
 	"golang.org/x/crypto/ripemd160"
 )
 
+type AddressEncodeDecoder struct {
+	AddressEncoder
+	AddressDecoder
+}
+
+type AddressEncoder struct {
+	params *Params
+}
+
+type AddressDecoder struct {
+	params *Params
+}
+
+func NewAddressEncoder(params *Params) AddressEncoder {
+	return AddressEncoder{params: params}
+}
+
+func NewAddressDecoder(params *Params) AddressDecoder {
+	return AddressDecoder{params: params}
+}
+
+func NewAddressEncodeDecoder(params *Params) AddressEncodeDecoder {
+	return AddressEncodeDecoder{
+		AddressEncoder: NewAddressEncoder(params),
+		AddressDecoder: NewAddressDecoder(params),
+	}
+}
+
+func (encoder AddressEncoder) EncodeAddress(rawAddr address.RawAddress) (address.Address, error) {
+	if len(rawAddr) != 26 && len(rawAddr) != 25 {
+		return address.Address(""), fmt.Errorf("address of unknown length")
+	}
+
+	var addrType uint8
+	var err error
+	var hash [20]byte
+	var prefix []byte
+	if len(rawAddr) == 26 {
+		prefix = rawAddr[:2]
+		addrType, _, err = parsePrefix(prefix)
+		copy(hash[:], rawAddr[2:22])
+	} else {
+		prefix = rawAddr[:1]
+		addrType, _, err = parsePrefix(prefix)
+		copy(hash[:], rawAddr[1:21])
+	}
+	if err != nil {
+		return address.Address(""), fmt.Errorf("parsing prefix: %v", err)
+	}
+
+	switch addrType {
+	case 0, 1: // P2PKH or P2SH
+		return address.Address(pack.String(encodeAddress(hash[:], prefix))), nil
+	default:
+		return address.Address(""), errors.New("unknown address")
+	}
+}
+
+func (decoder AddressDecoder) DecodeAddress(addr address.Address) (address.RawAddress, error) {
```

Comment:
What was the reason for copying/pasting the `decodeAddress` implementation into this method, instead of just calling `decodeAddress`? We should avoid copying/pasting where possible.

Reply:
Now we only have a single `DecodeAddress` implementing the Multichain Address API. Another method `addressFromRawBytes` lets us construct a `PubKeyHashAddress` or `ScriptAddress` struct, both of which implement the `bitcoincash.Address`/`zcash.Address` interfaces:
```go
func addressFromRawBytes(addrBytes []byte) (Address, error)
```